service/history/workflowExecutionContext.go

```diff
@@ -927,3 +932,35 @@ func (c *workflowExecutionContextImpl) emitSessionUpdateStats(stats *persistence
 	c.metricsClient.RecordTimer(metrics.SessionCountStatsScope, metrics.DeleteRequestCancelInfoCount,
 		time.Duration(stats.DeleteRequestCancelInfoCount))
 }
+
+func (c *workflowExecutionContextImpl) validateNoEventsAfterWorkflowFinish(input []*workflow.HistoryEvent) error {
+	if len(input) == 0 {
+		return nil
+	}
+
+	seenWorkflowCompletionEvent := false
+	for _, event := range input {
```

Comment:
instead of scan through entire list of events, how about we validate using the workflow state in mutableState.  Basically if workflow state is CLOSED and last event in the update is not one of 6 workflow close events then fail the transaction.

Reply:
i think there is an edge case about transient decision.
transient decision -> workflow completion will result to writing 2 batch of events to DB.
the first batch should be decision task scheduled & started (ms builder at this time indicate workflow closed), this batch will fail.